src/lib.rs

```diff
@@ -236,49 +193,59 @@ mod tests {
         ];
-        let mir2 = Mir::new(DefId::new(4, 5), blocks2);
-
-        let index = Index::new(2);
-        let mut curs = get_curs();
+        let mir2 = MetaData::Mir(Mir::new(DefId::new(4, 5), blocks2));
 
-        {
-            let mut enc = Encoder::new(&mut curs, index.clone()).unwrap();
-            enc.write_mir(mir1.clone()).unwrap();
-            enc.write_mir(mir2.clone()).unwrap();
-        }
-        rewind_curs(&mut curs);
+        vec![mir1, mir2]
+    }
 
-        let (got_index, mut dec) = Decoder::new(&mut curs).unwrap();
-        assert_eq!(got_index, index);
+    // Check serialising and deserialising works for zero MIRs.
+    #[test]
+    fn test_empty() {
+        let mut curs = get_curs();
 
-        let got_mir1 = dec.read_mir().unwrap();
-        assert_eq!(got_mir1, mir1);
+        let enc = Encoder::from(&mut curs).unwrap();
+        enc.done().unwrap();
 
-        let got_mir2 = dec.read_mir().unwrap();
-        assert_eq!(got_mir2, mir2);
+        rewind_curs(&mut curs);
+        let dec = Decoder::new(&mut curs).unwrap();
+        assert!(dec.iter().next().unwrap().is_none());
     }
 
-    // Try to decode more MIRs than we said we would.
+    // Check a typical serialising and deserialising session.
     #[test]
-    #[should_panic]
-    fn test_decode_too_many_mirs() {
-        let index = Index::new(2);
+    fn test_basic() {
+        let inputs = get_sample_metadata();
         let mut curs = get_curs();
 
-        let mut enc = Encoder::new(&mut curs, index.clone()).unwrap();
-        let mir = Mir::new(DefId::new(1, 2), Vec::new());
+        let mut enc = Encoder::from(&mut curs).unwrap();
+        for md in &inputs {
+            enc.serialise(md.clone()).unwrap();
+        }
+        enc.done().unwrap();
+
+        rewind_curs(&mut curs);
+        let dec = Decoder::new(&mut curs).unwrap();
+
+        // Obtain two fallible iterators, so we can zip them.
+        let got_iter = dec.iter();
+        let expect_iter = fallible_iterator::convert(inputs.into_iter().map(|e| Ok(e)));
 
-        enc.write_mir(mir.clone()).unwrap();
-        enc.write_mir(mir.clone()).unwrap();
-        enc.write_mir(mir).unwrap(); // panic.
+        let mut itr = got_iter.zip(expect_iter);
+        while let Some((got, expect)) = itr.next().unwrap() {
+            assert_eq!(expect, got);
+        }
     }
 
-    // Check that we get a panic if the encoder is still expecting more data to serialise.
     #[test]
-    #[should_panic]
-    fn test_drop_enc_premature() {
-        let index = Index::new(2);
+    #[should_panic(expected = "not marked done")]
+    fn test_encode_not_done() {
+        let inputs = get_sample_metadata();
         let mut curs = get_curs();
-        Encoder::new(&mut curs, index).unwrap();
+
+        let mut enc = Encoder::from(&mut curs).unwrap();
+        for md in &inputs {
+            enc.serialise(md.clone()).unwrap();
+        }
+        // User forgot: enc.done()
```

Comment:
This is a bit cryptic!

Reply:
Would this be better?
```
// The user forgot to finalise the serialisation with a call to `enc.done()`.
```

Reply:
You need to say something like "This test fails because the user forgot to add enc.done", otherwise the comment floats adrift in a universe of gnomisms.

Reply:
OK. I would have thought this was evident from `#[should_panic(expected = "not marked done")]`.

Reply:
I know your default style of comments, so I got it, but I doubt many other people would get it TBH.